packages/sdk/src/clients/optimisticOracleV2/client.ts

```diff
@@ -0,0 +1,185 @@
+import {
+  OptimisticOracleV2InterfaceEthers,
+  OptimisticOracleV2InterfaceEthers__factory,
+  getOptimisticOracleV2InterfaceAbi,
+} from "@uma/contracts-node";
+import type { SignerOrProvider, GetEventType } from "../..";
+import { Event, BigNumberish, utils, BigNumber } from "ethers";
+
+export type Instance = OptimisticOracleV2InterfaceEthers;
+const Factory = OptimisticOracleV2InterfaceEthers__factory;
+
+export function connect(address: string, provider: SignerOrProvider): Instance {
+  return Factory.connect(address, provider);
+}
+
+export const contractInterface = new utils.Interface(getOptimisticOracleV2InterfaceAbi());
+
+export type RequestPrice = GetEventType<Instance, "RequestPrice">;
+export type ProposePrice = GetEventType<Instance, "ProposePrice">;
+export type DisputePrice = GetEventType<Instance, "DisputePrice">;
+export type Settle = GetEventType<Instance, "Settle">;
+
+export enum RequestState {
+  Invalid = 0, // Never requested.
+  Requested, // Requested, no other actions taken.
+  Proposed, // Proposed, but not expired or disputed yet.
+  Expired, // Proposed, not disputed, past liveness.
+  Disputed, // Disputed, but no DVM price returned yet.
+  Resolved, // Disputed and DVM price is available.
+  Settled, // Final price has been set in the contract (can get here from Expired or Resolved).
+}
+
+// all events have these values
+export type RequestKey = {
+  requester: string;
+  identifier: string;
+  timestamp: number;
+  ancillaryData: string;
+};
+export type Request = RequestKey &
+  // this is partial since we dont know what events we have to populate parts of this
+  Partial<{
+    proposer: string;
+    disputer: string;
+    currency: string;
+    settled: boolean;
+    refundOnDispute: boolean;
+    proposedPrice: BigNumber;
+    resolvedPrice: BigNumber;
+    expirationTime: BigNumber;
+    reward: BigNumber;
+    finalFee: BigNumber;
+    bond: BigNumber;
+    customLiveness: BigNumber;
+    price: BigNumber;
+    payout: BigNumber;
+    state: RequestState;
+    // metadata about the transaction that triggered the state changes
+    requestTx: string;
+    proposeTx: string;
+    disputeTx: string;
+    settleTx: string;
+    requestBlockNumber: number;
+    proposeBlockNumber: number;
+    disputeBlockNumber: number;
+    settleBlockNumber: number;
+  }>;
+
+export interface EventState {
+  requests?: Record<string, Request>;
+}
+
+export function requestId(request: Omit<RequestKey, "timestamp"> & { timestamp: BigNumberish }): string {
+  // if enabling sorting, put timestamp first
+  return [request.timestamp.toString(), request.identifier, request.requester, request.ancillaryData].join("!");
+}
+
+export function reduceEvents(state: EventState, event: Event): EventState {
```

Comment:
If this file/code is identical to the regular OO, just with different type inputs (because it has a slightly different interface and state values), instead of copying this code, would it make more sense to just make these functions generic and specify the `T extends { ... }` where the `...` is just the required fields and use it for both? Or do we benefit from having completely distinct files/functions for each?

This is different than the skinny OO, since the skinny OO had far greater differences from the regular OO.

This may be something we could just add/improve in the future, too. Just wanted to get your thoughts.

Reply:
ye and the diff between the V1 and V2 OO is really just one small interface change on `RequestSettings` and the addition of some callback modifications that should not make a meaningful change to most parts of the flows.

Reply:
> If this file/code is identical to the regular OO, just with different type inputs (because it has a slightly different interface and state values), instead of copying this code, would it make more sense to just make these functions generic and specify the T extends { ... } where the ... is just the required fields and use it for both? Or do we benefit from having completely distinct files/functions for each?

When initially adding v2, i was uncertain as to what would need to be changed, so i started by just copying the code to see what would break.  It turns out nothing needed to be changed, so it will be possible to refactor this with generics, though i'd like to do that in a separate PR if thats ok.

Reply:
SGTM